overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+// seen, returns an error.
+func expandPathPatternFromIndex(pattern string, index, patternLen int) (expanded []string, end int, err error) {
+	// Record total list of expanded patterns, to which other lists are appended
+	expanded = []string{}
+	// Within the current group option, record the current list of previously-
+	// expanded prefixes, and the start index of the subpattern following the
+	// most recent group.
+	currPrefixes := defaultPrefixes
+	currSubpatternStart := index
+	for i := index; i < patternLen; i++ {
+		if pattern[i] == '\\' {
+			i += 1
+			if i == patternLen {
+				return nil, 0, fmt.Errorf(`invalid path pattern: trailing unescaped '\' character: %q`, pattern)
+			}
+			continue
+		}
+		if pattern[i] == '{' {
+			infix := pattern[currSubpatternStart:i]
+			groupExpanded, groupEnd, err := expandPathPatternFromIndex(pattern, i+1, patternLen)
+			if err != nil {
+				return nil, 0, err
+			}
+			newPrefixes := make([]string, 0, len(currPrefixes)*len(groupExpanded))
+			for _, prefix := range currPrefixes {
+				for _, suffix := range groupExpanded {
+					newPrefixes = append(newPrefixes, prefix+infix+suffix)
+				}
+			}
+			currPrefixes = newPrefixes
+			currSubpatternStart = groupEnd + 1
+			i = groupEnd // let for loop increment to index after '}'
+			continue
+		}
+		if pattern[i] == ',' || pattern[i] == '}' {
+			suffix := pattern[currSubpatternStart:i]
+			newExpanded := make([]string, len(expanded), len(expanded)+len(currPrefixes))
+			copy(newExpanded, expanded)
+			for _, prefix := range currPrefixes {
+				newExpanded = append(newExpanded, prefix+suffix)
+			}
+			expanded = newExpanded
+			currPrefixes = defaultPrefixes
+			currSubpatternStart = i + 1
+		}
+		if pattern[i] == '}' {
+			return expanded, i, nil
+		}
+	}
+	return nil, 0, fmt.Errorf(`invalid path pattern: unmatched '{' character: %q`, pattern)
+}
+
+var (
+	duplicateSlashes    = regexp.MustCompile(`(^|[^\\])/+`)
+	charsDoublestar     = regexp.MustCompile(`([^/\\])\*\*+`)
+	doublestarChars     = regexp.MustCompile(`([^\\])\*\*+([^/])`)
+	duplicateDoublestar = regexp.MustCompile(`/\*\*(/\*\*)+`) // relies on charsDoublestar running first
+	starsAnyMaybeStars  = regexp.MustCompile(`([^\\])\*+(\?\**)+`)
+)
+
+func cleanPattern(pattern string) string {
+	pattern = duplicateSlashes.ReplaceAllString(pattern, `${1}/`)
+	pattern = charsDoublestar.ReplaceAllString(pattern, `${1}*`)
+	pattern = doublestarChars.ReplaceAllString(pattern, `${1}*${2}`)
+	pattern = duplicateDoublestar.ReplaceAllString(pattern, `/**`)
+	pattern = starsAnyMaybeStars.ReplaceAllStringFunc(pattern, shiftStarAfterAnyChars)
+	if strings.HasSuffix(pattern, "/**/*") {
+		// Strip trailing "/*" from suffix
+		return pattern[:len(pattern)-2]
+	}
+	return pattern
+}
+
+func shiftStarAfterAnyChars(s string) string {
+	return strings.Map(deleteStars, s) + "*"
+}
+
+func deleteStars(r rune) rune {
+	if r == '*' {
+		return -1
+	}
+	return r
+}
```

Comment:
This is rather inefficient. I think @TheSignPainter98 mentioned some way to do a series of regex matches/substitutions on a single pass over a string. Any ideas how this could be improved?

Reply:
This may be possible with a regex set (something like Rust's [`RegexSet`](https://docs.rs/regex/latest/regex/struct.RegexSet.html)), but you'd need knowledge of which regex matched, where and what was the length of the match. I can't seem to see these in Go, but iirc as this is a one-time cost, perhaps multiple passes with a few regexes isn't too bad? Personally I'd prefer a loop though